helpers/target.go

```diff
@@ -129,3 +154,384 @@ func isAllowed(addr string) (bool, error) {
+	return ""
+}
+func (c *GitCreds) Username() string {
+	return c.User
+}
+func (c *GitCreds) Password() string {
+	return c.Pass
+}
+
+// IsReachable returns wether target is reachable
+// so the check execution can be performed.
+// If asset type is void, it will be inferred from target.
+//
+// ServiceCredentials are required for AWS, Docker and Git types.
+// Constructors for AWS, Docker and Git credentials can be found
+// in this same package.
+//
+// Verifications made depend on the asset type:
+//    - IP: None.
+//    - IPRange: None.
+//    - Hostname: NS Lookup resolution.
+//    - WebAddress: HTTP GET request.
+//    - DomainName: NS Lookup checking SOA record.
+//    - AWSAccount: Assume Role.
+//    - DockerImage: Docker pull.
+//    - GitRepository: Git ls-remote.
+//
+// This function does not return any output related to the process in order to
+// verify the target's reachability. This output can be useful for some cases
+// in order to not repeat work in the check execution (e.g.: Obtaining the
+// Assume Role token). For this purpose other individual methods can be called
+// from this same package with further options for AWS, Docker and Git types.
+func IsReachable(target, assetType string, creds ServiceCreds) (bool, error) {
+	var isReachable bool
+	var err error
+
+	// In order to support backward compatibility
+	// we have to support assetType being void.
+	if assetType == "" {
+		assetType, err = detectAssetType(target)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	if (assetType == awsAccType || assetType == dockerImgType ||
+		assetType == gitRepoType) && creds == nil {
+		return false, fmt.Errorf("ServiceCredentials are required")
+	}
+
+	switch assetType {
+	case hostnameType:
+		isReachable = IsHostnameReachable(target)
+	case webAddrsType:
+		isReachable = IsWebAddrsReachable(target)
+	case domainType:
+		isReachable, err = IsDomainReachable(target)
+	case awsAccType:
+		isReachable, _, err = IsAWSAccReachable(target, creds.URL(), creds.Username(), minSessTime)
+	case dockerImgType:
+		isReachable, err = IsDockerImgReachable(target, creds.Username(), creds.Password())
+	case gitRepoType:
+		isReachable = IsGitRepoReachable(target, creds.Username(), creds.Password())
+	default:
+		// Return true if we don't have a
+		// verification in place for asset type.
+		isReachable = true
+	}
+
+	return isReachable, err
+}
+
+func detectAssetType(target string) (string, error) {
+	if types.IsAWSARN(target) {
+		return awsAccType, nil
+	}
+	if types.IsDockerImage(target) {
+		return dockerImgType, nil
+	}
+	if types.IsGitRepository(target) {
+		return gitRepoType, nil
+	}
+	if types.IsIP(target) {
+		return ipType, nil
+	}
+	if types.IsCIDR(target) {
+		return ipRangeType, nil
+	}
+	if types.IsWebAddress(target) {
+		return webAddrsType, nil
+	}
+	if types.IsHostname(target) {
+		return hostnameType, nil
+	}
+	if isDomain, _ := types.IsDomainName(target); isDomain {
+		return domainType, nil
+	}
```

Comment:
Yes, I was aware, but I think there is no way to get this right for these cases on which the asset is a Hostname and DomainName, because independently of how we set the "priority" for the detection (checking first Domain or Hostname) you will get Hostname or DomainName, but is impossible to know if that was the intended type for that concrete checktype. Does it make sense?

Also, for these cases, the verification for the `IsReachable` would be the same as the detection, so all in all does not make much sense for theses cases.
Having this into account I think it would be better, as mentioned in the checks PR, to not check reachability for "backward compatible cases" on which we get a void `assetType`. What do you think?

Reply:
Asset detection has been deprecated, see comments below regarding backward compatiblity.